eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/DualFundingHandlers.scala

```diff
@@ -59,25 +59,34 @@ trait DualFundingHandlers extends CommonFundingHandlers {
     }
   }
 
+  def pruneCommitments(w: WatchFundingConfirmedTriggered, d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED): Try[DualFundingTx] = {
+    val allFundingTxs: Seq[DualFundingTx] = DualFundingTx(d.fundingTx, d.commitments) +: d.previousFundingTxs
+    // We can forget other funding attempts now that one of the funding txs is confirmed.
+    val otherFundingTxs = allFundingTxs.filter(_.commitments.fundingTxId != w.tx.txid).map(_.fundingTx)
+    rollbackDualFundingTxs(otherFundingTxs)
+    // We find which funding transaction got confirmed.
+    allFundingTxs.find(_.commitments.fundingTxId == w.tx.txid) match {
+      case Some(dft: DualFundingTx) =>
+        log.info("channelId={} was confirmed at blockHeight={} txIndex={} with funding txid={}", d.channelId, w.blockHeight, w.txIndex, w.tx.txid)
+        watchFundingTx(dft.commitments)
+        context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, w.tx))
+        Success(dft)
+      case None =>
+        val t = new RuntimeException(s"internal error: the funding tx that confirmed doesn't match any of our funding txs: ${w.tx.bin}")
+        Failure(t)
+    }
+  }
+
+  /**
+   * We could just ignore the event and handle it after the next reconnection, but this allows us to eagerly rollback
+   * deprecated funding transactions
+   */
   def handleDualFundingConfirmedOffline(w: WatchFundingConfirmedTriggered, d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED) = {
-    if (w.tx.txid == d.commitments.fundingTxId) {
-      watchFundingTx(d.commitments)
-      context.system.eventStream.publish(TransactionConfirmed(d.channelId, remoteNodeId, w.tx))
-      // We can forget previous funding attempts now that the funding tx is confirmed.
-      rollbackDualFundingTxs(d.previousFundingTxs.map(_.fundingTx))
-      stay() using d.copy(previousFundingTxs = Nil) storing()
-    } else if (d.previousFundingTxs.exists(_.commitments.fundingTxId == w.tx.txid)) {
-      log.info("channelId={} was confirmed at blockHeight={} txIndex={} with a previous funding txid={}", d.channelId, w.blockHeight, w.txIndex, w.tx.txid)
```

Comment:
There was a reason why that function only added a log line when a previous transaction confirmed, this ensured the log line wouldn't be repeated every time you restart eclair if the peer doesn't reconnect. This can be annoying for Phoenix swaps for example, which could cause logs from many channels to pollute the logs after every restart...

It's easy to fix though:

- remove the `log.info` from `pruneCommitments`
- move it to the event handler of `ChannelOpenDualFunded`
- in `handleDualFundingConfirmedOffline`, in the `Success` case, we can compare the confirmed txId with what we previously had in `d`, and if they don't match add the log line that says a previous funding tx confirmed

Reply:
You are referring to the watch set here:
https://github.com/ACINQ/eclair/blob/7b9a4a15539913c43e6f7c23fb01e09861612843/eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala#L320-L322

Shouldn't we only set it when there are multiple competing funding txs instead?

Reply:
We have to set those watches to make sure we detect offline force-closes and correctly get our main output back...

Reply:
Right, this is where we put the `WatchFundingSpent`, gotcha.

Reply:
I went with a slight variation of your suggestion in [d92c894](https://github.com/ACINQ/eclair/pull/2557/commits/d92c894384a8378221966e72b9bac8a62145ad9d).

In `handleDualFundingConfirmedOffline`, I believe that we are really interested in the case where we are actuzlly pruning _any_ funding tx, rather than a previous vs latest.